tokio-util/src/sync/mpsc.rs

```diff
@@ -1,220 +1,274 @@
+    /// 
+    /// # Panics
+    /// 
+    /// If a previous call to `poll_reserve` returned `Poll::Ready(Ok(()))`, and `poll_reserve` is
+    /// called again immediately after without a call to `start_send` first, the second call to
+    /// `poll_reserve` will panic.
+    pub fn poll_reserve(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), PollSendError<T>>> {
+        let (result, next_state) = match self.take_state() {
+            State::Idle(sender) => {
+                self.acquire.set(make_acquire_future(Some(sender)));
+                match self.acquire.poll(cx) {
+                    // Channel has capacity.
+                    Poll::Ready(Ok(permit)) => (Poll::Ready(Ok(())), State::ReadyToSend(permit)),
+                    // Channel is closed.
+                    Poll::Ready(Err(e)) => (Poll::Ready(Err(e)), State::Closed),
+                    // Channel doesn't have capacity yet, so we need to wait.
+                    Poll::Pending => (Poll::Pending, State::Acquiring),
+                }
+            },
+            State::Acquiring => match ready!(self.acquire.poll(cx)) {
+                // Channel has capacity.
+                Ok(permit) => (Poll::Ready(Ok(())), State::ReadyToSend(permit)),
+                // Channel is closed.
+                Err(e) => (Poll::Ready(Err(e)), State::Closed), 
+            },
+            // We're closed, either by choice or because the underlying sender was closed.
+            State::Closed => return Poll::Ready(Err(PollSendError(None))),
+            State::ReadyToSend(_) => panic!("called `poll_ready` when `start_send` was expected"),
+        };
 
-        let result = self.inner.poll(cx);
-        if result.is_ready() {
-            self.is_sending = false;
-        }
-        if let Poll::Ready(Err(_)) = &result {
-            self.sender = None;
-        }
+        self.state = next_state;
         result
     }
 
-    /// Check whether the channel is ready to send more messages now.
-    ///
-    /// If this method returns `true`, then `start_send` is guaranteed to not
-    /// panic.
-    ///
-    /// If the channel is closed, this method returns `true`.
-    pub fn is_ready(&self) -> bool {
-        !self.is_sending
+    /// Sends an item to the channel.
+    /// 
+    /// Before calling `start_send`, `poll_reserve` must be called with a successful return
+    /// value of `Poll::Ready(Ok(()))`.
+    /// 
+    /// # Errors
+    /// 
+    /// If the channel is closed, an error will be returned.  This is a permanent state.
+    /// 
+    /// # Panics
+    /// 
+    /// If `poll_reserve` was not successfully called prior to calling `start_send`, then this method
+    /// will panic.
+    pub fn start_send(&mut self, value: T) -> Result<(), PollSendError<T>> {
+        let (result, next_state) = match self.take_state() {
+            State::Idle(_) | State::Acquiring => panic!("`start_send` called without first calling `poll_ready`"),
```

Comment:
We could make the choice to have `start_send` succeed in this case, attempting to use the `send` method instead?

Reply:
The only viable approach I currently see to make this method more resilient is that we could _attempt_ a send via `Sender::try_send`.  We'd then be able to change the language in the documentation to something more like:

> If `poll_reserve` is called prior to calling `start_send`, and returns `Poll::Ready(Ok(()))`, then the call to `start_send` is guaranteed to succeed.  If it is not called prior, then a send is attempted but may or may not succeed.

The biggest issue then would be how to document the error case: `PollSendError<T>` is meant to only ever indicate that the channel is closed, but in this case, `try_send` might fail due to the channel being closed _or_ full.  It feels weird to differentiate errors based on the condition of whether or not `poll_reserve` was called since there is supposed to be a formal contract of calling `poll_reserve` _before_ `start_send`.

Long story short, I'm willing to make such a change, just curious what your thoughts are on how to best document the behavior difference depending on if `poll_reserve` was called first or not.